File: frontend/lib/api/projects.ts
/**
 * API client for project management endpoints.
 *
 * Handles project CRUD operations and summary retrieval.
 */

import { apiRequest } from "@/lib/api-client";

export interface Project {
  project_id: string;
  name: string;
  description?: string;
  tags: string[];
  repository_count: number;
  created_at: string;
  updated_at: string;
  last_scan_at?: string | null;
  active: boolean;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  tags?: string[];
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  tags?: string[];
  active?: boolean;
}

export interface ProjectSummary {
  project_id: string;
  project_name: string;
  repository_count: number;
  total_scans: number;
  findings: {
    critical: number;
    high: number;
    medium: number;
    low: number;
    total: number;
  };
  top_vulnerabilities: Array<Record<string, any>>;
  compliance_status: Record<string, any> | null;
  repositories: Array<{
    repository_id: string;
    repository_name: string;
    scan_count: number;
    last_scan_at: string | null;
  }>;
}

/**
 * Create a new project.
 */
export async function createProject(
  data: CreateProjectRequest
): Promise<{ success: boolean; data: Project }> {
  return apiRequest<{ success: boolean; data: Project }>(`/v1/projects`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * List all projects for the organization.
 */
export async function listProjects(params?: {
  tags?: string;
  active?: boolean;
  limit?: number;
  offset?: number;
}): Promise<{ success: boolean; data: { projects: Project[]; total: number } }> {
  const queryParams = new URLSearchParams();
  if (params?.tags) queryParams.append("tags", params.tags);
  if (params?.active !== undefined) queryParams.append("active", params.active.toString());
  if (params?.limit) queryParams.append("limit", params.limit.toString());
  if (params?.offset) queryParams.append("offset", params.offset.toString());

  const qs = queryParams.toString();
  return apiRequest(`/v1/projects${qs ? `?${qs}` : ""}`);
}

/**
 * Get project details by ID.
 */
export async function getProject(
  projectId: string
): Promise<{ success: boolean; data: Project }> {
  return apiRequest(`/v1/projects/${projectId}`);
}

/**
 * Update a project.
 */
export async function updateProject(
  projectId: string,
  data: UpdateProjectRequest
): Promise<{ success: boolean; data: Project }> {
  return apiRequest(`/v1/projects/${projectId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * Soft delete a project.
 */
export async function deleteProject(
  projectId: string
): Promise<{ success: boolean; data: { project_id: string; name: string; deleted_at: string; message: string } }> {
  return apiRequest(`/v1/projects/${projectId}`, {
    method: "DELETE",
  });
}

/**
 * Get project summary with aggregated stats.
 */
export async function getProjectSummary(
  projectId: string
): Promise<{ success: boolean; data: ProjectSummary }> {
  return apiRequest(`/v1/projects/${projectId}/summary`);
}

File: src/api/models/responses/projects.py
"""
Project response models.
"""

from pydantic import BaseModel, Field
from typing import Optional, List


class ProjectResponse(BaseModel):
    """Response model for project details."""

    project_id: str = Field(..., description="Unique project identifier")
    name: str = Field(..., description="Project name")
    description: Optional[str] = Field(None, description="Project description")
    tags: List[str] = Field(default_factory=list, description="Project tags")
    repository_count: int = Field(0, description="Number of repositories in this project")
    created_at: str = Field(..., description="Creation timestamp (ISO 8601)")
    updated_at: str = Field(..., description="Last update timestamp (ISO 8601)")
    last_scan_at: Optional[str] = Field(None, description="Most recent scan timestamp across all repositories")
    active: bool = Field(True, description="Active status (false = soft deleted)")


class CreateProjectResponse(BaseModel):
    """Response when creating a new project."""

    project_id: str = Field(..., description="Unique project identifier")
    name: str = Field(..., description="Project name")
    description: Optional[str] = Field(None, description="Project description")
    tags: List[str] = Field(default_factory=list, description="Project tags")
    created_at: str = Field(..., description="Creation timestamp (ISO 8601)")
    repository_count: int = Field(0, description="Number of repositories (always 0 at creation)")


class ListProjectsResponse(BaseModel):
    """Response for listing projects."""

    projects: List[ProjectResponse] = Field(..., description="List of projects")
    total: int = Field(..., description="Total number of projects")


class ProjectSummaryResponse(BaseModel):
    """Response for project-level aggregation/summary."""

    project_id: str = Field(..., description="Project identifier")
    project_name: str = Field(..., description="Project name")
    repository_count: int = Field(..., description="Number of repositories in project")
    total_scans: int = Field(..., description="Total scan sessions across all repositories")
    findings: dict = Field(..., description="Aggregated findings breakdown")
    top_vulnerabilities: List[dict] = Field(default_factory=list, description="Top CVEs across project")
    compliance_status: Optional[dict] = Field(None, description="Compliance status summary")
    repositories: List[dict] = Field(default_factory=list, description="Repository breakdown with findings")


class DeleteProjectResponse(BaseModel):
    """Response when deleting (soft delete) a project."""

    project_id: str = Field(..., description="Deleted project identifier")
    name: str = Field(..., description="Project name")
    deleted_at: str = Field(..., description="Deletion timestamp (ISO 8601)")
    message: str = Field(..., description="Deletion confirmation message")
